### PR TITLE
Gallery: Add labels to img, figure and figcaption elements for accessibility

### LINKED ISSUE #13164: Linked gallery images are not accessible
**Describe the bug**
On the front-end (published page), a gallery block with linked images is not accessible. Screen readers cannot read the images in a gallery if they are linked. 

* All of the linked images in the gallery are listed by the image's file name in the "Links" list. (Tested in Voiceover on Mac).
* All of the linked images in the gallery are read by the screen reader as just "link". (Tested in NVDA on Windows.)

This is happening because the `<a>` tags in the gallery have not ARIA label or text content. The caption of each image is in the `<figure>` tag, but outside of the `<a>` tag:

```
<figure>
  <a href="URL-OF-IMAGE.JPG">
    <img src="URL-OF-IMAGE.JPG">
  </a>

  <figcaption>CAPTION TEXT</figcaption>
</figure>
```

Either an `aria-label` attribute needs to be added to the `` tag or a `title` attribute needs to be added to the `<a>` tag.

**To Reproduce**
1. Go to a page that contains a gallery block with linked images.
2. Read the page with a screen reader, specifically the gallery block.

**Expected behavior**
The links inside the gallery should be read as understandable text, not the image's URL or "link".

### PR DESCRIPTION
**Pull request by scruffian**
Fixes #13164

## Description
This adds some attributes to the Gallery block markup for accessibility reasons:
- Add the `aria-describedby` attribute to the `figure` element to declare what element contains the caption
- Add an `id` to the corresponding `figcaption` to associate it with the figure
- Also add the caption to the `aria-label` attribute on the image, as a fallback in case the describedby attribute doesn't work
- Also add the caption to the `title` attribute on the link, as a fallback in case the describedby attribute doesn't work

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Review by enriquesanchez — APPROVED** [review]
Tested with VoiceOver on Safari and tried various combinations of images with and without `alt`, and with and without caption.

In all cases I felt enough information was being announced 👍

File: packages/block-library/src/gallery/gallery-image.js [hunk 1]
```diff
@@ -281,6 +292,7 @@ class GalleryImage extends Component {
 						}
 						unstableOnFocus={ this.onSelectCaption }
 						inlineToolbar
+						id={ 'caption' + id }
```
Comment (ZebulanStanphill) [inline packages/block-library/src/gallery/gallery-image.js:295]: There's a typo here:
```suggestion
						id={ 'caption-' + id }
```

File: packages/block-library/src/gallery/gallery-image.js [hunk 2]
```diff
@@ -213,6 +213,7 @@ class GalleryImage extends Component {
 					tabIndex="0"
 					aria-label={ ariaLabel }
 					ref={ this.bindContainer }
+					aria-describedby={ 'caption-' + id }
```
Comment (ZebulanStanphill) [inline packages/block-library/src/gallery/gallery-image.js:216]: Our [coding guidelines](https://developer.wordpress.org/block-editor/contributors/develop/coding-guidelines/#strings) say to use template literals whenever possible.

```suggestion
					aria-describedby={ `caption-${ id }` }
```

**Review by ZebulanStanphill — DISMISSED** [review]

**Comment by tellthemachines** [comment]
Tested this PR with both VoiceOver/Safari and NVDA/Firefox and found that `aria-describedby` doesn't stop the URL from being read out, and on NVDA/Firefox it doesn't read out the caption at all. 

I've pushed a commit replacing it with a conditional `aria-label`, which will render the image caption ONLY if the image has no alt text defined, if there is a caption and if the image is inside a link. Testing on both the above screen readers shows this solution to work well.

**Comment by enriquesanchez** [comment]
Thanks, @tellthemachines!

Tested on VoiceOver and this is what I found:

**1. If the image is a link and only has a caption**
The caption gets announced twice:

Screen Shot 2020-09-25 at 16 04 52


**2. If the image is a link and only has alt text**
The alt text gets announced as expected:

Screen Shot 2020-09-25 at 15 56 16


**3. If the image is a link and has both alt text and a caption**
The alt text and caption get announced:

Screen Shot 2020-09-25 at 15 56 56

**4. If the image is a link and has no alt text nor a caption**
The image's URL gets announced:

Screen Shot 2020-09-25 at 15 57 55

---

I think this is feeling in good shape with the exception of #1. I don't think the caption should be announced twice. What do you think?

**Comment by tellthemachines** [comment]
> I think this is feeling in good shape with the exception of #1. I don't think the caption should be announced twice. What do you think?

Hmm, I tried adding `aria-hidden` to the `figcaption` when the image has an `aria-label`, and that fixes the repetition in VoiceOver, but unfortunately causes a repetition in NVDA 😖. As it is, NVDA announces "figure [caption text] graphic link", but with `aria-hidden` on the `figcaption` it announces "[caption text] figure [caption text] graphic link". 

I can't think of any other way to avoid the repetition in VoiceOver, and having to choose between the two, I'd rather optimise for NVDA as it's the more popular screen reader. 
I'm also not super-keen on adding `aria-hidden` because then screen reader users may not be able to check that a caption has been output. What are your thoughts @enriquesanchez ?

**Comment by enriquesanchez** [comment]
@tellthemachines I agree, definitively not a blocker and much better than what we had before. Thank you!

**Comment by ZebulanStanphill** [comment]
I'm not able to review at the moment, but I'd like to dismiss my previous review. How do I do that? I can't find any button to do so.

**Review by noisysocks — APPROVED** [review]
Code looks good and the block migration worked in my testing 👍

**Comment by mtias** [comment]
Did we consider adding these server side instead of serializing?

**Comment by afercia** [comment]
Thanks for this PR!

Did we consider to not use `aria-label` at all? When there's no alt text and there is a caption, the capiton could be used as the alt text. While the implementation in this PR works for software that understand ARIA, it doesn't work when the image is broken, not loaded, or images are disabled (because of reasons). For clarity, see the screenshot below:

On the left: broken image with no alt text and with aria-label
On the right: broken image with the aria caption value used as alt text

Screenshot 2020-10-01 at 15 56 46

The alt text is always preferable as it can help also sighted users. Also, we should consider [the first rule of ARIA use](https://www.w3.org/TR/using-aria/#firstrule). /Cc @joedolson

**Comment by enriquesanchez** [comment]
>When there's no alt text and there is a caption, the capiton could be used as the alt text.

@afercia I did not think of that scenario, thanks for raising it. Wouldn't the presence of the caption be enough?

**Comment by mtias** [comment]
If we aim to revert this can we do it before 9.2 is released to avoid extra deprecations?

**Comment by afercia** [comment]
> I did not think of that scenario, thanks for raising it. Wouldn't the presence of the caption be enough?

@enriquesanchez my point is more about why to use an aria-label in that scenario when it's possible to populate the alt text? Usage of ARIA should be avoided when there's an equivalent HTML element or attribute that can be used.

**Comment by ZebulanStanphill** [comment]
@afercia I've made #26082 to improve the behavior.

**Pull request merged**